make visual mode edit more intuitive

pkfm opened the pull request:
adds `<Plug>(ScalpelVisualSelection)` which yanks the visual selection into scalpel's pattern field. also adds `<Plug>(ScalpelVisualSmart)` which mixes `ScalpelVisual` and `ScalpelVisualSelection` behaviour, using the latter in most cases, and the former when the selection is only one 'non-special' character.

this behavior was mentioned in [issue #3](https://github.com/wincent/scalpel/issues/3).

auto selection of the word under the cursor doesn't make sense in visual mode as the point of visual mode is visual selection of text, which is unnecessary when selection is automatic (or rather a matter of positioning) as it is with the current visual mode behavior. why would you move to visual mode and begin a selection only to disregard the selection?

this behavior might make sense in a two stage process where you first select the range, and then select the word, but then a range is probably not of great use with scalpel always invoking interactive replacement. ([chrisbra/NrrwRgn](https://github.com/chrisbra/NrrwRgn) is a plugin that accomplishes this)

the scalpel precision analogy makes more sense with the behavior here as visual mode allows you to make a more precise selection over and/or within the word boundary. consider some example selections. `'[...]'` denotes visual selection.

```
some 'selection' text    " original
some '[select]ion' text  " select root word
some ['selection'] text  " include surrounding characters
some ['sel]ection' text  " select 'prefix' and opening punctation
[some 'selection'] text  " select across whitespace
aCamel[Case]Var          " select tag components
a_[snake_case]_var       " select tag components
```

this only supports selections within a single line. selections beyond line breaks would be useful, especially when performing substitutions on comments, but this is what i've got so far.

the 'smart' variant assumes selection is deliberate unless its a single character. if the character is loosely a valid variable/method name character (w.r.t. many languages), then it falls back on `scalpel#cword` i.e. the original behavior. this provides an easy way to change basically 'non-word' symbols, like for example, changing several strings with single quotes to strings with double quotes. here we can skip over the escaped quotes with scalpel, which would be equivalent to `:%s/'/"/gc`, but it's nice to be able to do everything with the same (`<leader>e`) motion.

'foo bar'
'foo\'s bar'
'bar's foo'
'bar foo'

this also provides a way to get both behaviors in most cases from a single map with a little user awareness. an option like `g:ScalpelVisualSmartFallbackChars` might be in order if this all seems reasonable to you.

sorry this is more wordy than the actual diff and thank you for this valuable plugin.

pkfm committed:
make visual mode edit more intuitive

adds <Plug>(ScalpelVisualSelection) which yanks the visual selection into scalpel's pattern field. also adds <Plug>(ScalpelVisualSmart) which mixes ScalpelVisual and ScalpelVisualSelection behaviour, using the latter in most cases, and the former when the selection is only one 'non-special' chararacter.

wincent commented on plugin/scalpel.vim:
```diff
@@ -337,13 +344,21 @@ execute 'vnoremap <Plug>(ScalpelVisual) :' .
   endif
-  if !hasmapto('<Plug>(ScalpelVisual)') && maparg('<leader>e', 'v') ==# ''
-    vmap <unique> <Leader>e <Plug>(ScalpelVisual)
+  if !hasmapto('<Plug>(ScalpelVisualSmart)') && maparg('<leader>e', 'v') ==# ''
+    vmap <unique> <Leader>e <Plug>(ScalpelVisualSmart)
```
Because of the way people tend to install Vim plugins (ie. with a plug-in manager, pointing at `HEAD` of `master`), we can't really _ever_ change behavior in breaking ways. And that includes breaking it when we bump the major version number, because people never see the version number, and therefore never have the opportunity to decline/avoid the breakage.

So that means it is OK to _add_ functionality (like the new `ScalpelVisualSmart` and ScalpelVisualSelection`), but not OK to overwrite previously existing functionality with it.

That is, you can pick a different mapping for it, or not map it at all — and instead mention it in the docs. People who are interested are then free to assign it to `<leader>e` if they want.

wincent reviewed (commented):
Thanks for the PR @pkfm.

The truth is, the VISUAL mode support was something I added because I could, but it's simply not all that useful. I probably use Scalpel 10 to 100 times a day, but I use it in VISUAL mode about 1 in 1,000 of those times. I think the behavior you're proposing here is definitely more useful.

I'm not entirely convinced of the `ScalpelVisualSmart` behavior — it seems a bit arbitrary. But I guess I'll have to take it for a test-drive to see if it "feels" right/useful.

Other than that, we'll need some docs for the new mappings before we can merge this. Do you want to take a shot at adding those? At the moment, because there is only one mapping, I just have it documented in the flow of the help document and not in a dedicated "mappings" section. But as this adds a couple more mappings, you'd have to use the `@mapping` annotation ([example from another project here](https://github.com/wincent/ferret/blob/6218891a01418377c67b12ad57046ec78958ced2/plugin/ferret.vim#L773-L784)) — and we can generate the README.md and the help doc itself from the code comments.

pkfm committed:
Reverts visual mode map

pkfm committed:
Fixes typo in previous commit

pkfm committed:
Documents new mappings

Reorganizes some of the doc text to into configuration and mappings
section to accomodate new mappings that aren't mapped by default.
Additionally adds inline mapping annotations to vim source files.
Annotations on unmapped functinality and differences between inline and
dedicated docs may be problematic

pkfm committed:
Sync 'header' doc text in plugin/scalpel.vim with previous commit

The update is little more than a copy+paste of doc/scalpel.txt contents
to plugin/scalpel.vim, with some comment commands and a global replace
to change section commands from upper to lower case.

pkfm committed:
Comments empty lines in doc section to match previous style

pkfm committed:
Removes all traces of <Plug>(ScalpelVisualSmart) map

Hopefully.

pkfm commented:
woops. I didn't mean to push a remap. It should be reverted now. I agree that the smart mode isn't very useful and I haven't used it much since the initial PR so I removed it. I also made an attempt at documentation. I'm not sure what default map to suggest, and since there is only one additional command I'm also not sure that it is necessary to use `@mapping` annotations. I also couldn't find anything about them in the docs, so I went ahead and added them in addition to amending the doc text, which is slightly re-organized for config/maps sections. lmk if that is appropriate or not at all what you meant.

pkfm commented:
I just now realized you've written a documentation generator. I suppose the proposed `<Plug>` command should use an `@command` annotation instead.

wincent commented:
Yeah, don't worry about the doc generator (it's a pain to install); wherever you want to stick the docs I can just copy-paste them into the right place and massage them until they look right.
